packages/wdio-cli/src/launcher.js

```diff
@@ -53,6 +53,27 @@ class Launcher {
         this.runnerFailed = 0
     }
 
+    /**
+     * Combinate test data from data provider and retries with specs
+     */
+    getSpecsWithDataProviderAndRetries(config, capSpecs, capExclude) {
+        /**
+         * avoid retries in watch mode
+         */
+        const specFileRetries = config.watch ? 0 : config.specFileRetries
+        let specs = []
+        const dataProvidersMap = this.configParser.getDataProviders() || {}
+
+        this.configParser.getSpecs(capSpecs, capExclude).forEach((specFile) => {
+            let specDataSet = dataProvidersMap[specFile] || ['']
+            specDataSet.forEach((data) => {
+                specs.push({ files: [specFile], testData: data, retries: specFileRetries })
+            })
+        })
```

Comment:
It seems like nothing really happens in the loop if `dataProvidersMap` is an empty object. Can't we return an empty `specs` array if `this.configParser.getDataProviders()` is null?

Example:

```js
const dataProvidersMap = this.configParser.getDataProviders() || {}

if (dataProvidersMap) {
  // do loop here
}

return specs

```